clang/lib/Sema/SemaAvailability.cpp

```diff
@@ -157,6 +157,18 @@ static bool ShouldDiagnoseAvailabilityInContext(
     }
   }
 
+  // In HLSL, emit diagnostic during parsing only if the diagnostic
+  // mode is set to strict (-fhlsl-strict-availability), and either the decl
+  // availability is not restricted to a specific environment/shader stage,
+  // or the target stage is known (= it is not shader library).
+  if (S.getLangOpts().HLSL) {
+    if (!S.getLangOpts().HLSLStrictAvailability ||
+        (DeclEnv != nullptr &&
+         S.getASTContext().getTargetInfo().getTriple().getEnvironment() ==
```

Comment:
Here (and a bit further down in the PR as well) there are places that assume that "not library" implies "known target stage". I wonder if we're opening ourselves up to subtle bugs issues in the future if we end up adding another library-like target?

Maybe this is unlikely enough that we can deal with that later?

Reply:
Yes, that is the current assumption everywhere - it is either library or a shader with a known target stage. Unless you see any potential library-like target on the horizon I think we can leave it like that.

Reply:
@llvm-beanz mentioned something this morning that made me think there may be a possibility of another environment that isn't a library shader, but also isn't tied to a specific shader stage.  I may have misunderstood though.

If we've already got a load of code making this same assumption then maybe now isn't the right time to tackle it anyway. Happy for no change here.